### PR TITLE
Sprint 3

### PR DESCRIPTION
**Pull request by MarkoNik**
https://github.com/RAF-SI-2023/Banka-1-Backend/issues/120

Implemented web scraper for futures fetching from yahoo finance.

### CODE REVIEW

File: market-service/src/main/java/rs/edu/raf/banka1/mapper/FutureMapper.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import rs.edu.raf.banka1.model.ListingFuture;
+import rs.edu.raf.banka1.model.ListingStock;
+import rs.edu.raf.banka1.model.OptionsModel;
+import rs.edu.raf.banka1.model.dtos.ListingBaseDto;
+import rs.edu.raf.banka1.model.dtos.ListingForexDto;
+import rs.edu.raf.banka1.model.dtos.ListingFutureDto;
+import rs.edu.raf.banka1.model.dtos.OptionsDto;
+@Component
+public class FutureMapper {
+    public void updateFuture(ListingFuture oldFuture, ListingFuture newFuture) {
```
Comment (sudo0rw3ll) [inline market-service/src/main/java/rs/edu/raf/banka1/mapper/FutureMapper.java:14]: I think that it's better to move functions which are not related to mapper functionality (transforming DTOs to Model and Model to DTO) to a separate file.

File: market-service/src/main/java/rs/edu/raf/banka1/model/ListingFuture.java [hunk 2]
```diff
@@ -17,4 +17,6 @@ public class ListingFuture extends ListingBase {
     private Integer openInterest;
     @Column
     private Integer settlementDate;
+    @Column
+    private String alternativeTicker;
```
Comment (ijevtic) [inline market-service/src/main/java/rs/edu/raf/banka1/model/ListingFuture.java:21]: za sta je ovo potrebno?

File: market-service/src/main/java/rs/edu/raf/banka1/services/FuturesService.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import java.util.Optional;
+
+public interface FuturesService {
+    List<ListingFuture> fetchNFutures(int n);
+
+    List<ListingHistory> fetchNFutureHistories(List<ListingFuture> listingFutures, int n);
+
+    int addAllFutures(List<ListingFuture> futures);
+
+    int addFuture(ListingFuture future);
```
Comment (ijevtic) [inline market-service/src/main/java/rs/edu/raf/banka1/services/FuturesService.java:18]: zasto se vraca int?

File: market-service/src/main/java/rs/edu/raf/banka1/services/FuturesServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+        ListingFuture future = futureRepository.findById(id).orElse(null);
+        if(future == null){
+            return listingHistories;
+        }
+
+        String ticker = future.getTicker();
+        listingHistories = listingHistoryRepository.getListingHistoriesByTicker(ticker);
+        if(listingHistories.isEmpty()) {
+            WebDriver driver = new ChromeDriver(options);
+            listingHistories = fetchNSingleFutureHistory(future, 20, driver);
```
Comment (ijevtic) [inline market-service/src/main/java/rs/edu/raf/banka1/services/FuturesServiceImpl.java:303]: jel mozemo izvuci ovo u constants fajl za laksi tuning kasnije

File: market-service/src/main/java/rs/edu/raf/banka1/threads/FutureThread.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    private final ListingStockService listingStockService;
+
+    public FutureThread(FuturesService futuresService, ListingStockService listingStockService) {
+        this.futuresService = futuresService;
+        this.listingStockService = listingStockService;
+    }
+
+    @Override
+    public void run() {
+        var futures = futuresService.fetchNFutures(10);
```
Comment (ijevtic) [inline market-service/src/main/java/rs/edu/raf/banka1/threads/FutureThread.java:19]: i ovo u constants

File: market-service/src/main/java/rs/edu/raf/banka1/threads/FutureThread.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        futuresService.addAllFutures(futures);
+        var histories = futuresService.fetchNFutureHistories(futures, 20);
```
Comment (ijevtic) [inline market-service/src/main/java/rs/edu/raf/banka1/threads/FutureThread.java:21]: var?

File: market-service/src/test/java/rs/edu/raf/banka1/services/FuturesServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
+package rs.edu.raf.banka1.services;
```
Comment (ijevtic) [inline market-service/src/test/java/rs/edu/raf/banka1/services/FuturesServiceImplTest.java:1]: @savatarko

**Review by sudo0rw3ll — APPROVED** [review]
Everything seems okay. Checked scraping logic. I would just write a general note which is related to scraping. It's always the best to select elements by their ID or CLASS_NAME if ID is not present. 

Those things are less possible to change inside HTML.
There are also CSS_SELECTORS and X_PATHS as selection methods.
X_PATH is the last resort. If there are no element IDs, CLASS_NAMES, CSS_SELECTORS then this is used. It's the worst option since X_PATH gives you the position of the element itself inside the HTML (DOM TREE). If we add simple paragraph inside the  HTML, X_PATH of all elements are changed and if scrapers are not maintained then they'll fail.

Sorry for this, but I just felt like sharing some things with you guys.

File: market-service/src/main/java/rs/edu/raf/banka1/threads/FutureThread.java (hunk 6 above)
Comment (sudo0rw3ll) [reply market-service/src/main/java/rs/edu/raf/banka1/threads/FutureThread.java:21]: I meni je bilo cudno ali zapravo moze da se koristi xD

File: market-service/src/main/java/rs/edu/raf/banka1/mapper/FutureMapper.java (hunk 1 above)
Comment (MarkoNik) [reply market-service/src/main/java/rs/edu/raf/banka1/mapper/FutureMapper.java:14]: made it related to mapper by returning an object

**Review by savatarko — APPROVED** [review]

**Pull request merged**